include/powerloader/target.hpp

```diff
@@ -41,47 +41,182 @@ namespace powerloader
+        }
+
+        // TODO: don't expose ownership
+        std::shared_ptr<Mirror>& mirror() noexcept
+        {
+            assert(m_mirror);
+            return m_mirror;
+        }
+
+        void change_mirror(std::shared_ptr<Mirror> mirror);
+
+        const fs::path temp_file() const noexcept
+        {
+            return m_temp_file;
+        }
+
+        bool writecb_required_range_written() const noexcept
+        {
+            return m_writecb_required_range_written;
+        }
+
+        HeaderCbState headercb_state() const noexcept
+        {
+            return m_headercb_state;
+        }
+
+        const std::string& headercb_interrupt_reason() const noexcept
+        {
+            return m_headercb_interrupt_reason;
+        }
+
+        int range_fail() const noexcept
+        {
+            return m_range_fail;
+        }
+
+        void reset_range_fail() noexcept
+        {
+            m_range_fail = false;
+        }
+
+        std::size_t retries() const noexcept
+        {
+            return m_retries;
+        }
+
+        DownloadState state() const noexcept
+        {
+            return m_state;
+        }
+
+        ZckState zck_state() const noexcept
+        {
+            return m_zck_state;
+        }
+
+        // TODO: refactor to avoid state being changed directly from outisde.
+        void set_zck_state(ZckState new_state) noexcept
+        {
+            m_zck_state = new_state;
+        }
+
+        const auto& errorbuffer() const noexcept
+        {
+            return m_errorbuffer;
+        }
+
+        const auto& mirrors() const noexcept
+        {
+            return m_mirrors;
+        }
+
+        const auto& tried_mirrors() const noexcept
+        {
+            return m_tried_mirrors;
+        }
+
+        CURLHandle* curl_handle() const noexcept
+        {
+            return m_curl_handle.get();
+        }
+
+    private:
+        std::shared_ptr<DownloadTarget> m_target;
+        fs::path m_temp_file;
+        std::string m_url_stub;
+
+        bool m_resume = false;
+        std::size_t m_resume_count = 0;
+        std::ptrdiff_t m_original_offset;
 
         // internal stuff
-        std::size_t retries = 0;
+        std::size_t m_retries = 0;
 
-        DownloadState state = DownloadState::kWAITING;
+        DownloadState m_state = DownloadState::kWAITING;
 
         // mirror list (or should we have a failure callback)
-        std::shared_ptr<Mirror> mirror = nullptr;
-        std::vector<std::shared_ptr<Mirror>> mirrors;
-        std::set<std::shared_ptr<Mirror>> tried_mirrors;
-        std::shared_ptr<Mirror> used_mirror = nullptr;
-
-        HeaderCbState headercb_state;
-        std::string headercb_interrupt_reason;
-        std::size_t writecb_received;
-        bool writecb_required_range_written;
+        std::shared_ptr<Mirror> m_mirror;
+        std::vector<std::shared_ptr<Mirror>> m_mirrors;
+        std::set<std::shared_ptr<Mirror>> m_tried_mirrors;
```

Comment:
We should typedef these types so it's easier and more expressive to use them

Reply:
OK, I'll improve this as part of the ownership changes as the types will change.